Bluetooth: ascs: Handle ASE control point operations in separate thread

MariuszSkamra opened the pull request:
This adds handling of ASE control point operations in separate thread
so that the notifications of ASE state changes are sent from non-BTRX
thread. This ensures bt_gatt_notify_cb to be blocking waiting for
available buffers to send the notifications.

MariuszSkamra commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -2810,6 +2828,76 @@ static ssize_t ascs_cp_write(struct bt_conn *conn,
+	}
+}
+
+NET_BUF_POOL_DEFINE(cp_write_buf_pool, ASCS_MAX_CP_WRITES, BT_ATT_BUF_SIZE, 0, NULL);
```
As the Control Point write operation can be GATT Long Write procedure, the size of this buffer is not sufficient to handle it.

Thalley commented:
@MariuszSkamra Before I dive too deep into the implementation of this, what is the advantage of using a new thread for this compared to the system workqueue thread?

larsgk commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -2905,4 +2993,7 @@ void bt_ascs_cleanup(void)
 	}
 }
+
+K_THREAD_DEFINE(ascs, CONFIG_BT_ASCS_STACK_SIZE, ascs_thread, NULL, NULL, NULL,
```
I am wondering if we can get this working without the need for a separate thread (thinking memory constraints)?

MariuszSkamra replied:
I decided to make use of the System Workqueue thread

larsgk commented:
> This adds handling of ASE control point operations in separate thread so that the notifications of ASE state changes are sent from non-BTRX thread. This ensures bt_gatt_notify_cb to be blocking waiting for available buffers to send the notifications.

Being slightly cautious adding to the memory budget, I am wondering if this can be done without a separate thread.  Also, do we have some real life tests/observations where this causes a clear issue (sans thread)?

MariuszSkamra committed:
Bluetooth: audio: ascs: Defer ASE state transition

This adds handling of ASE control point operations in separate thread
so that the notifications of ASE state changes are sent from non-BT
thread. This ensures bt_gatt_notify_cb to be blocking waiting for
available buffers to send the notifications.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -236,37 +241,145 @@ static int ascs_disconnect_stream(struct bt_bap_stream *stream)
 
-	if (ep->stream == NULL) {
-		return;
+	if (stream->ops != NULL && stream->ops->configured != NULL) {
+		stream->ops->configured(stream, &ase->ep.qos_pref);
 	}
+}
```
General comment: Consider storing `stream->ops` in a local variable to reduce number of dereferences (not for performance, but rather just code readability).

Applies multiple places

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -285,198 +398,143 @@ void ascs_ep_set_state(struct bt_bap_ep *ep, uint8_t state)
-				ops->released(stream);
-			}
-			struct bt_ascs_ase *ase = CONTAINER_OF(ep, struct bt_ascs_ase, ep);
+	switch (new_state) {
+	case BT_BAP_EP_STATE_IDLE:
+		ase_set_state_idle(ase);
+		break;
+	case BT_BAP_EP_STATE_CODEC_CONFIGURED:
+		ase_set_state_codec_configured(ase);
+		break;
+	case BT_BAP_EP_STATE_QOS_CONFIGURED:
+		ase_set_state_qos_configured(ase);
+		break;
+	case BT_BAP_EP_STATE_ENABLING:
+		ase_set_state_enabling(ase);
+		break;
+	case BT_BAP_EP_STATE_STREAMING:
+		ase_set_state_streaming(ase);
+		break;
+	case BT_BAP_EP_STATE_DISABLING:
+		ase_set_state_disabling(ase);
+		break;
+	case BT_BAP_EP_STATE_RELEASING:
+		ase_set_state_releasing(ase);
+		break;
+	default:
+		__ASSERT_PRINT("Invalid state %d", new_state);
+	}
```
This does make it significantly easier to read - We should consider applying the same for the other BAP state machines

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -285,198 +398,143 @@ void ascs_ep_set_state(struct bt_bap_ep *ep, uint8_t state)
+		case BT_BAP_EP_STATE_RELEASING:
+			break;
+		default:
+			goto invalid_transition;
```
Could consider using a local `bool` as `bool valid_state_transition` and then have a `if (valid_state_transition)` at the end of the function to avoid disrupting the flow of the function.

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -285,198 +398,143 @@ void ascs_ep_set_state(struct bt_bap_ep *ep, uint8_t state)
+	ase->state_pending = state;
 
-			ep->receiver_ready = false;
+	err = k_work_submit(&ase->state_transition_work);
+	if (err < 0) {
+		return err;
```
Maybe we should log this? 

If this indeed fails, then we can't really recover from it, can we?

Thalley replied:
Indeed. We can't really do much besides logging this as an `LOG_ERR`; I don't think we can do anything to recover from it

Thalley reviewed (dismissed):
Changes are sound, and none of my comments are blocking

carlescufi commented:
@MariuszSkamra please rebase

MariuszSkamra committed:
Bluetooth: audio: ascs: Remove unused variable

The `ops` variable is unused, thus can be removed.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>
